### PR TITLE
docs: parallelization

### PR DESCRIPTION
**Pull request by cristianmtr**

### CODE REVIEW

File: chapters/parallel/index.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+##### Index vs Query time Polling
+
+We set `polling` to `any` in order to tell the `HeadPea` to spread the data across its inner `Peas`. This allows us to parallelize. You do this for the Flow you will use to index data. 
+
+However, when querying, you need to make sure to use a Flow where the `Pods` are using `polling: all`. This ensures that the requests are seen by all the `Peas`, and that the respective id is found. If we use `any` we risk that the key is not seen by the shard that has the data. 
+
+[comment]: <> (TODO)
+
+### Limitations [optional]
```
Comment (FionnD) [inline chapters/parallel/index.md]: Remove please since you are not using it

**Review by FionnD — CHANGES_REQUESTED** [review]

File: chapters/parallel/index.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
+## Parallelization in Jina
+
+##### Learn about how Jina allows for parallelization of data processing
```
Comment (alexcg1) [inline chapters/parallel/index.md]: ~about~
~for~

File: chapters/parallel/index.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+[comment]: <> (1. Parallel pods)
+
+[comment]: <> (2. Parallel peas within pod)
+
+[comment]: <> (2a. sharding)
+
+[comment]: <> (2b. parallel processing)
+
+Jina allows for parallelization of Peas and Pods. This generally leads to an increase in throughput, as the data is split across processes. However, depending on the configuration, this can also lead to
```
Comment (alexcg1) [inline chapters/parallel/index.md]: lead to...what?

File: chapters/parallel/index.md [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+[comment]: <> (2a. sharding)
+
+[comment]: <> (2b. parallel processing)
+
+Jina allows for parallelization of Peas and Pods. This generally leads to an increase in throughput, as the data is split across processes. However, depending on the configuration, this can also lead to 
+
+### Before you start
+
+You are expected to be familiar with the general concepts of Jina. You should also have experience with parallelism in other systems. Jina does follow the best practices and designs from other databases or search engines. However, Jina is primarily a *neural search engine*. This means some changes and adaptations were necessary. This chapter will explain these.
```
Comment (alexcg1) [inline chapters/parallel/index.md]: - `general concepts of Jina` - link to docs overview page

File: chapters/parallel/index.md [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+The plot of the above topology looks like this:
+
+![](./simple-plot3.svg)
+
+In the above, `p1`, `p2`, `p3` now subscribe to `gateway` and conduct their work in parallel. The last `.needs()` blocks all Pods until they finish their work.
+
+#### Intra-parallelism
+
+Peas are the next level at which we can configure parallelism. In this case, `Pod` receives the data (in its `HeadPea`) and then distributes it to its `Peas`. You can achieve this by using the `parallel` parameter:
```
Comment (alexcg1) [inline chapters/parallel/index.md:38]: Unbacktick anything that isn't text you see in CLI input/output or code (e.g. Pod, Peas)

@FionnD Do we have guidance on this in style guide? I see a lot of inconsistency and room for confusion (to be honest, with myself also)

File: chapters/parallel/index.md [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+           .add(name='p2', needs='gateway')
+           .add(name='p3', parallel=3)
+           .needs(['p1','p3'], name='r1').plot())
+```
+
+![](./simple-plot4.svg)
+
+#### State-less vs state-ful
+
+Depending on the type of `Peas`, `parallel` means different things. In *state-less* `Peas` (Crafters, Encoders etc.), you *parallelize*, not *shard*. This means the data is not persisted, but it's only processed in a parallel manner.
```
Comment (alexcg1) [inline chapters/parallel/index.md]: Unbacktick again (here and through the rest of the doc)

**Review by alexcg1 — CHANGES_REQUESTED** [review]

File: chapters/parallel/index.md (hunk 5 above)
Comment (cristianmtr) [reply chapters/parallel/index.md:38]: https://developers.google.com/style/text-formatting says code formatting should be added for class names (Peas) and variables (parallel)

Comment (alexcg1) [reply chapters/parallel/index.md:38]: - `parallel` - no issue with that :smile: 
- Peas/Pods - I thought you were talking about the things themselves, not the class names. If talking class names, keep the backticks and ignore my prior comment!

File: chapters/parallel/index.md (hunk 4 above)
Comment (cristianmtr) [reply chapters/parallel/index.md]: Which one is that?

**Review by alexcg1 — APPROVED** [review]

File: chapters/parallel/index.md (hunk 3 above)
Comment (cristianmtr) [reply chapters/parallel/index.md]: Forgot what I was going to write :thinking: Couldn't have been important then. Removed

File: chapters/parallel/index.md (hunk 2 above)
Comment (cristianmtr) [reply chapters/parallel/index.md]: thanks

File: chapters/parallel/index.md (hunk 4 above)
Comment (alexcg1) [reply chapters/parallel/index.md]: Good question! @FionnD do we have a placeholder link for that section yet?

File: chapters/parallel/index.md [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+Depending on the type of `Peas`, `parallel` means different things. In *state-less* `Peas` (`Crafters`, `Encoders` etc.), you *parallelize*, not *shard*. This means the data is not persisted, but it's only processed in a parallel manner. 
+
+On the other hand, *state-ful* `Peas` (Indexers) will store the data across *shards*. This terminology is borrowed from other database systems
+
+For more on sharding in other DB systems (with a focus on traditional SQL), see [here](https://subscription.packtpub.com/book/big_data_and_business_intelligence/9781849516723/1/ch01lvl1sec10/using-sharding-and-data-distribution)
+
+For both of these, you need to set `parallel` to some number higher than `1`, and `polling` to `any`.
+
+##### Index vs Query time Polling
```
Comment (JoanFM) [inline chapters/parallel/index.md:60]: I think you can add some words about `scheduling` configuration for the sake of completness. What do you think?

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: chapters/parallel/index.md (hunk 7 above)
Comment (cristianmtr) [reply chapters/parallel/index.md:60]: Done

File: chapters/parallel/index.md [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+However, when querying, you need to make sure to use a Flow where the `Pods` are using `polling: all`. This ensures that the requests are seen by all the `Peas`, and that the respective `id` is found. If we use `polling: any` we risk that the key is not seen by the shard that has the data.
+
+#### Scheduling
+
+Jina offers two types of scheduling:
+
+1. `round_robin` will assign tasks in circular order, without any prioritization. We assume all `Peas` have uniform processing power.
+
+2. `load_balance` will assign tasks to the faster `Peas`.
```
Comment (JoanFM) [inline chapters/parallel/index.md]: that is not 100% accurate. It will balance the load to 'idle' peas

**Review by JoanFM — DISMISSED** [review]

File: chapters/parallel/index.md (hunk 8 above)
Comment (cristianmtr) [reply chapters/parallel/index.md]: Done

**Review by alexcg1 — APPROVED** [review]

File: chapters/parallel/index.md [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+           .needs(['p1','p3'], name='r1').plot())
+```
+
+![](./simple-plot4.svg)
+
+#### State-less vs state-ful
+
+Depending on the type of `Peas`, `parallel` means different things. In *state-less* `Peas` (`Crafters`, `Encoders` etc.), you *parallelize*, not *shard*. This means the data is not persisted, but it's only processed in a parallel manner. 
+
+On the other hand, *state-ful* `Peas` (Indexers) will store the data across *shards*. This terminology is borrowed from other database systems
```
Comment (Yongxuanzhang) [inline chapters/parallel/index.md]:
```suggestion
On the other hand, *state-ful* `Peas` (Indexers) will store the data across *shards*. This terminology is borrowed from other database systems.
```

File: chapters/parallel/index.md [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+For more on sharding in other DB systems (with a focus on traditional SQL), see [here](https://subscription.packtpub.com/book/big_data_and_business_intelligence/9781849516723/1/ch01lvl1sec10/using-sharding-and-data-distribution)
```
Comment (Yongxuanzhang) [inline chapters/parallel/index.md]:
```suggestion
For more on sharding in other DB systems (with a focus on traditional SQL), see [here].(https://subscription.packtpub.com/book/big_data_and_business_intelligence/9781849516723/1/ch01lvl1sec10/using-sharding-and-data-distribution)
```

File: chapters/parallel/index.md [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+However, when querying, you need to make sure to use a Flow where the `Pods` are using `polling: all`. This ensures that the requests are seen by all the `Peas`, and that the respective `id` is found. If we use `polling: any` we risk that the key is not seen by the shard that has the data.
+
+#### Scheduling
+
+Jina offers two types of scheduling:
+
+1. `round_robin` will assign tasks in circular order, without any prioritization. We assume all `Peas` have uniform processing power.
+
+2. `load_balance` will assign tasks to the idle `Peas`.
+
```
Comment (Yongxuanzhang) [inline chapters/parallel/index.md:73]: Shall we add a code snippet on scheduling? Like
```python
from jina.enums import  SchedulerType
f = Flow().add(
        name='p1',
        parallel=10, scheduling=SchedulerType.ROUND_ROBIN)
```

Comment (JoanFM) [reply chapters/parallel/index.md:73]: that makes sense

Comment (cristianmtr) [reply chapters/parallel/index.md:73]: Good idea. Done

File: chapters/parallel/index.md (hunk 4 above)
Comment (FionnD) [reply chapters/parallel/index.md]: It will be 
chapters/core/introduction/index

**Review by FionnD — APPROVED** [review]
Looks great to me :)

File: chapters/parallel/index.md (hunk 4 above)
Comment (cristianmtr) [reply chapters/parallel/index.md]: Is the way I added it correct?

Comment (FionnD) [reply chapters/parallel/index.md]: Should be. I've taken a notice that I will personally check back and make sure it works once we merge the future PR.

**Comment by cristianmtr** [comment]
@bwanglzu Should I merge with the failing checks? Or wait?

**Comment by bwanglzu** [comment]
@cristianmtr merge it, the failing checks is still WIP, please ignore it at the moment.

**Pull request merged**